feat(Kafka): Add option to login to SASL password-protected Kafka, like MSK

Smejky338 opened the pull request.

Smejky338 committed:
feat: Add option to send to SASL-enabled clusters like MSK

Smejky338 committed:
feat(kafka getTimes): Add option to connect to SASL-enabled clusters like MSK

Smejky338 committed:
fix(HBI_populate): Add support for multiple brokers in arg

Smejky338 committed:
feat(HBI_utils): Add separate producer for password-protected Kafka

Smejky338 committed:
feat: remove duplicate console output

Smejky338 committed:
fix(hbi_populate): Wrap in try/catch for attributeException

Smejky338 committed:
fix(hbi_utils): Fix order of password vs. nonpassword KafkaProducer

Smejky338 committed:
fix(Kafka args): Add Kafka args in both places, don't default to a value

Smejky338 commented:
In the first use, I got these timeout errors so that only a part of the hosts generated:
https://master-jenkins-csb-perf.apps.ocp-c1.prod.psi.redhat.com/job/InsightsROS_runner/397/console
`2023-12-08 15:44:12,709 kafka.client WARNING <BrokerConnection node_id=3 host=b-3-public.openpocmsk.pyzafp.c6.kafka.us-east-2.amazonaws.com:9196 <connected> [IPv4 ('3.132.12.218', 9196)]> timed out after 30000 ms. Closing connection.
2023-12-08 15:44:12,709 kafka.client WARNING Node 3 connection failed -- refreshing metadata
2023-12-08 15:44:12,710 kafka.producer.record_accumulator WARNING Produced messages to topic-partition TopicPartition(topic='platform.inventory.host-ingress', partition=8) with base offset -1 log start offset None and error None.
2023-12-08 15:44:12,710 root ERROR Failed to send message (RequestTimedOutError('Request timed out after 30000 ms'),); {'data_stats': {'successes': 7, 'failures': 0}, 'data_lock': <unlocked _thread.lock object at 0x7f44f173ccc0>}
2023-12-08 15:44:12,710 kafka.client WARNING <BrokerConnection node_id=2 host=b-2-public.openpocmsk.pyzafp.c6.kafka.us-east-2.amazonaws.com:9196 <connected> [IPv4 ('3.131.20.190', 9196)]> timed out after 30000 ms. Closing connection.
2023-12-08 15:44:12,710 kafka.producer.record_accumulator WARNING [Error 7] RequestTimedOutError: Request timed out after 30000 ms
2023-12-08 15:44:12,710 kafka.client WARNING Node 2 connection failed -- refreshing metadata
2023-12-08 15:44:12,711 kafka.producer.record_accumulator WARNING Produced messages to topic-partition TopicPartition(topic='platform.inventory.host-ingress', partition=1) with base offset -1 log start offset None and error None.
2023-12-08 15:44:12,711 root ERROR Failed to send message (RequestTimedOutError('Request timed out after 30000 ms'),); {'data_stats': {'successes': 7, 'failures': 1}, 'data_lock': <unlocked _thread.lock object at 0x7f44f173ccc0>}
2023-12-08 15:44:12,711 kafka.producer.record_accumulator WARNING Produced messages to topic-partition TopicPartition(topic='platform.inventory.host-ingress', partition=4) with base offset -1 log start offset None and error None.
2023-12-08 15:44:12,711 root ERROR Failed to send message (RequestTimedOutError('Request timed out after 30000 ms'),); {'data_stats': {'successes': 7, 'failures': 2}, 'data_lock': <unlocked _thread.lock object at 0x7f44f173ccc0>}
2023-12-08 15:44:12,711 kafka.producer.record_accumulator WARNING [Error 7] RequestTimedOutError: Request timed out after 30000 ms
2023-12-08 15:44:12,711 kafka.producer.record_accumulator WARNING [Error 7] RequestTimedOutError: Request timed out after 30000 ms
Traceback (most recent call last):
  File "/home/venv/bin/script-hbi-populate.py", line 33, in <module>
    sys.exit(load_entry_point('opl-rhcloud-perf-team', 'console_scripts', 'script-hbi-populate.py')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/venv/src/opl-rhcloud-perf-team/opl/hbi_utils.py", line 311, in populate_main
    gen_send_verify(args, status_data)
  File "/home/venv/src/opl-rhcloud-perf-team/opl/hbi_utils.py", line 217, in gen_send_verify
    gen_and_send(
  File "/home/venv/src/opl-rhcloud-perf-team/opl/hbi_utils.py", line 107, in gen_and_send
    raise Exception(
Exception: Failed to send 3 messages out of totally requested 10
command terminated with exit code 1`

... but the second one worked fine: https://master-jenkins-csb-perf.apps.ocp-c1.prod.psi.redhat.com/job/InsightsROS_runner/398/console

Smejky338 committed:
fix(Kafka): Switch to try/except AttributeError for switching of SASL x noauth

Smejky338 committed:
feat(kafka): unify using of kafka_host:port

Smejky338 committed:
fix linting Kafka_times

Smejky338 committed:
fix lint: avoid code duplication

Smejky338 committed:
lint fix - post_kafka

Smejky338 committed:
fix error message "existing"/"only have"

Smejky338 committed:
feat(skip_to_end): Add possibility to use SASL auth into Kafka - for MSK

Smejky338 committed:
feat(consumer_lag): Add possibility to use SASL auth into Kafka - for MSK

Smejky338 committed:
fix(skip_to_end): fix hash key: value to be a pair

Smejky338 committed:
fix: Subscribe to the topic after consumer creation, don't use the config hash as it doesn't work that way

shubham-html-css-js commented on opl/consumer_lag.py:
```diff
@@ -12,26 +12,52 @@ class ConsumerLag:
-        )
+        # Common parameters for both cases
+        common_params = {
+            topic: self.topic,
```
I think this "topic" key should be removed from here too

Smejky338 replied:
I see, it shouldn't subscribe there anyways as it uses this method: https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html?highlight=partitions_for_topic#kafka.KafkaConsumer.partitions_for_topic

shubham-html-css-js commented on opl/consumer_lag.py:
```diff
@@ -12,26 +12,52 @@ class ConsumerLag:
+            group_id: self.group,
+            session_timeout_ms: 50000,
+            heartbeat_interval_ms: 10000,
+            consumer_timeout_ms: 100000,
```
I think the keys should be wrapped in quotes

shubham-html-css-js commented on opl/skip_to_end.py:
```diff
@@ -20,20 +20,39 @@ def doit_seek_to_end(kafka_hosts, kafka_timeout, kafka_topic, kafka_group):
+        group_id: kafka_group,
+        session_timeout_ms: 50000,
+        heartbeat_interval_ms: 10000,
+        consumer_timeout_ms: kafka_timeout,
```
I think the keys should be wrapped in quotes

Smejky338 committed:
fix(consumer_lag): Remove topic from KafkaConsumer config

it uses partitions_for_topic, no sub

Smejky338 committed:
fix(kafka): missing doublequote wrap of hash keys

thanks @shubham-html-css-js

Smejky338 committed:
fix(consumer_lag): logging message vars

Smejky338 committed:
fix(consumer_lag): wrong timeout var name in logging

jhutar commented on opl/consumer_lag.py:
```diff
@@ -12,26 +12,50 @@ class ConsumerLag:
-            consumer_timeout_ms=100000,
-        )
+        # Common parameters for both cases
+        common_params = {
```
Nice!

jhutar commented on opl/consumer_lag.py:
```diff
@@ -12,26 +12,50 @@ class ConsumerLag:
+        }
 
+        # Kafka consumer creation: SASL or noauth
+        try:
```
With this try we will print first log info message every time. Could you replace the `try:` directly with `if self.username != "":`?

jhutar replied:
Or maybe `if self.username is not None:` given how default is specified in `add_kafka_opts(`)

Smejky338 replied:
Or maybe I should set the add_kafka_opts default to ""?

jhutar commented on opl/get_kafka_times.py:
```diff
@@ -78,26 +80,43 @@ def kafka_ts2dt(self, timestamp):
+            )
+            consumer = KafkaConsumer(
+                self.kafka_topic,
+                bootstrap_servers=self.kafka_host,  # self.kafka_hosts,
```
I think `bootstrap_servers` needs to be a list, not just one string.

Smejky338 replied:
It works even as a single string: https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html
This way should be more flexible to alter between single and many servers, but I guess the flexible way won't work without some additional code :)

jhutar commented on opl/get_kafka_times.py:
```diff
@@ -78,26 +80,43 @@ def kafka_ts2dt(self, timestamp):
-        logging.debug(
-            f"Created Kafka consumer for {self.kafka_hosts} for {self.kafka_topic} topic in group {self.kafka_group} with {self.kafka_timeout} ms timeout"
-        )
+        try:
```
Same here. Can you do a common list of options and only add sasl and security opts `if username is not None`?

jhutar commented on opl/hbi_utils.py:
```diff
@@ -188,10 +188,26 @@ def gen_send_verify(args, status_data):
     if args.dry_run:
         producer = None
     else:
+        try:
```
Again, please use if so we do not log `Creating SASL password-protected producer to {args.kafka_host}` everytime.

jhutar commented on opl/post_kafka_times.py:
```diff
@@ -328,18 +328,35 @@ def produce_thread(args, config, produce_here, save_here):
+            "compression_type": args.compression_type,
+        }
+
+        try:
```
Also here please.

jhutar commented on opl/skip_to_end.py:
```diff
@@ -20,20 +20,39 @@ def doit_seek_to_end(kafka_hosts, kafka_timeout, kafka_topic, kafka_group):
+    }
+
+    # Kafka consumer creation: SASL or noauth
+    try:
```
You know :)

jhutar commented on opl/skip_to_end.py:
```diff
@@ -57,6 +76,7 @@ def doit(args, status_data):
         args.kafka_timeout,
         args.kafka_topic,
         args.kafka_group,
+        args,
```
Change it just to send args then as we have all of the other stuff there anyway

Smejky338 replied:
Good point!

shubham-html-css-js commented:
@jsmejkal @jhutar I have tried this in my local environment and for Patchman , the script works well

Smejky338 committed:
feat: default kafka opts username&psw to ""

Smejky338 committed:
feat(kafka): avoid try block to use the default value "" of kafka vars

Smejky338 committed:
refactor(kafka consumers): Switch to common params hash and put consumer code into separate method, use default kafka username = ""

Smejky338 commented:
I've changed to default the kafka_user and kafka_password to "" and modified the scripts to take that into account and also refactored some Kafka consumer/producer code to be separated into a method.
It would be nice to refactor all the kafka consumer and producer methods into a common place so we don't have that much copypaste, but I would skip that "for now".
